fails to build with ghc-9.4 for Stackage Nightly

```
Building library for calligraphy-0.1.3..                          
[ 3 of 16] Compiling Calligraphy.Compat.Lib                                                                                         

/var/stackage/work/unpack-dir/unpacked/calligraphy-0.1.3-307771ebc155176f83b616559925c8fe11c75b7aac351cb58edb01a696dd4163/src/Callig
raphy/Compat/Lib.hs:48:38: error:
    Data constructor not in scope:                                
      NCU :: ((NameCache -> (NameCache, b0)) -> IO b0) -> NameCache                                                                 
   |                                                              
48 | readHieFileCompat ref = readHieFile (NCU (atomicModifyIORef ref))                                                              
   |                                      ^^^                                          
```

--- a/.github/workflows/CI.yaml
+++ b/.github/workflows/CI.yaml
@@ -5,26 +5,29 @@ on: push
 jobs:
 
   nix-build:
-    name: Nix default build
+    name: Nix default
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v3
       - uses: cachix/install-nix-action@v19
       - run: nix build -L
-      - run: nix build -L .#devShell.x86_64-linux
+      - run: nix develop -L -c echo "All good"
 
-  nix-shells:
-    name: Nix build ${{ matrix.ghc }} shell
+  nix-matrix:
+    name: Nix ${{ matrix.ghc }}
     strategy:
       matrix:
-        # TODO bump flake, ghc, add ghc94
-        ghc: [ ghc884, ghc8107, ghc902, ghc922 ]
+        ghc: [ ghc8107, ghc902, ghc927, ghc944 ]
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v3
       - uses: cachix/install-nix-action@v19
-      - run: nix build -L .#devShells.x86_64-linux.${{ matrix.ghc }}-shell
-      - run: nix develop -L .#devShells.x86_64-linux.${{ matrix.ghc }}-shell -c cabal new-build
+      - name: build shell
+        run: nix build -L .#devShells.x86_64-linux.${{ matrix.ghc }}-shell
+      - name: cabal build in shell
+        run: nix develop -L .#devShells.x86_64-linux.${{ matrix.ghc }}-shell -c cabal new-build
+      - name: nix build
+        run: nix build -L .#apps.x86_64-linux.calligraphy-${{ matrix.ghc }}
 
   stack-build:
     name: Stack ${{ matrix.resolver }} on ${{ matrix.os }}

--- a/.github/workflows/CI.yaml
+++ b/.github/workflows/CI.yaml
@@ -8,8 +8,8 @@ jobs:
     name: Nix default build
     runs-on: ubuntu-latest
     steps:
-      - uses: actions/checkout@v2
-      - uses: cachix/install-nix-action@v16
+      - uses: actions/checkout@v3
+      - uses: cachix/install-nix-action@v19
       - run: nix build -L
       - run: nix build -L .#devShell.x86_64-linux
 
@@ -21,8 +21,8 @@ jobs:
         ghc: [ ghc884, ghc8107, ghc902, ghc922 ]
     runs-on: ubuntu-latest
     steps:
-      - uses: actions/checkout@v2
-      - uses: cachix/install-nix-action@v16
+      - uses: actions/checkout@v3
+      - uses: cachix/install-nix-action@v19
       - run: nix build -L .#devShells.x86_64-linux.${{ matrix.ghc }}-shell
       - run: nix develop -L .#devShells.x86_64-linux.${{ matrix.ghc }}-shell -c cabal new-build
 
@@ -35,8 +35,8 @@ jobs:
         os: [ macos-latest, ubuntu-latest ]
     runs-on: ${{ matrix.os }}
     steps:
-      - uses: actions/checkout@v2
-      - uses: haskell/actions/setup@v1
+      - uses: actions/checkout@v3
+      - uses: haskell/actions/setup@v2
         with:
           enable-stack: true
       - run: stack init --resolver ${{ matrix.resolver }}
@@ -49,15 +49,15 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        ghc: [ '8.8.4', '8.10.7', '9.0.2', '9.2.2', '9.4.4' ] # TODO bump all to latest
+        ghc: [ '8.8.4', '8.10.7', '9.0.2', '9.2.7', '9.4.4' ]
         os: [ ubuntu-latest, macos-latest ]
     runs-on: ${{ matrix.os }}
     steps:
-      - uses: actions/checkout@v2
-      - uses: haskell/actions/setup@v1
+      - uses: actions/checkout@v3
+      - uses: haskell/actions/setup@v2
         with:
           ghc-version: ${{ matrix.ghc }}
-      - uses: actions/cache@v2
+      - uses: actions/cache@v3
         with:
           path: |
             ~/.cabal

--- a/.github/workflows/CI.yaml
+++ b/.github/workflows/CI.yaml
@@ -17,7 +17,8 @@ jobs:
     name: Nix build ${{ matrix.ghc }} shell
     strategy:
       matrix:
-        ghc: [ 'ghc884', 'ghc8107', 'ghc902', 'ghc922' ]
+        # TODO bump flake, ghc, add ghc94
+        ghc: [ ghc884, ghc8107, ghc902, ghc922 ]
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v2
@@ -30,7 +31,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        resolver: [ nightly, lts-19, lts-18, lts-17, lts-16, lts-15 ]
+        resolver: [ nightly, lts-20, lts-19, lts-18, lts-17, lts-16, lts-15 ]
         os: [ macos-latest, ubuntu-latest ]
     runs-on: ${{ matrix.os }}
     steps:
@@ -48,7 +49,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        ghc: [ '8.8.4', '8.10.7', '9.0.2', '9.2.2' ]
+        ghc: [ '8.8.4', '8.10.7', '9.0.2', '9.2.2', '9.4.4' ] # TODO bump all to latest
         os: [ ubuntu-latest, macos-latest ]
     runs-on: ${{ matrix.os }}
     steps:

--- a/flake.nix
+++ b/flake.nix
@@ -16,12 +16,23 @@
         calligraphy = final.haskell.lib.compose.justStaticExecutables final.haskellPackages.calligraphy;
       };
 
+      supported-ghc-versions = [
+        "ghc922"
+        "ghc902"
+        "ghc8107"
+        "ghc884"
+      ];
+      default-ghc-version = "ghc922";
+      per-compiler = fkey: fattr:
+        (builtins.listToAttrs (builtins.map (str: { name = fkey str; value = fattr str; }) supported-ghc-versions))
+        // { default = fattr default-ghc-version; };
+
       perSystem = system:
         let
           pkgs = import inputs.nixpkgs { inherit system; overlays = [ overlay ]; };
           mkApp = compiler:
-             pkgs.haskell.lib.compose.justStaticExecutables
-               pkgs.haskell.packages."${compiler}".calligraphy;
+            pkgs.haskell.lib.compose.justStaticExecutables
+              pkgs.haskell.packages."${compiler}".calligraphy;
           mkShell = compiler:
             let hspkgs = pkgs.haskell.packages.${compiler}; in
             hspkgs.shellFor {
@@ -38,23 +49,11 @@
               ];
             };
         in
-        rec {
-          devShells = {
-            ghc922-shell = mkShell "ghc922";
-            ghc902-shell = mkShell "ghc902";
-            ghc8107-shell = mkShell "ghc8107";
-            ghc884-shell = mkShell "ghc884";
-          };
-          devShell = devShells.ghc922-shell;
+        {
+          devShells = per-compiler (str: str + "-shell") mkShell;
           packages.calligraphy = pkgs.calligraphy;
           defaultPackage = pkgs.calligraphy;
-
-          apps = {
-            calligraphy-ghc922 = mkApp "ghc922";
-            calligraphy-ghc902 = mkApp "ghc902";
-            calligraphy-ghc8107 = mkApp "ghc8107";
-            calligraphy-ghc884 = mkApp "ghc884";
-          };
+          apps = per-compiler (str: "calligraphy-" + str) mkApp;
         };
     in
     { inherit overlay; } // inputs.flake-utils.lib.eachDefaultSystem perSystem;

--- a/flake.nix
+++ b/flake.nix
@@ -17,12 +17,13 @@
       };
 
       supported-ghc-versions = [
-        "ghc922"
+        "ghc944"
+        "ghc927"
         "ghc902"
         "ghc8107"
         "ghc884"
       ];
-      default-ghc-version = "ghc922";
+      default-ghc-version = "ghc944";
       per-compiler = fkey: fattr:
         (builtins.listToAttrs (builtins.map (str: { name = fkey str; value = fattr str; }) supported-ghc-versions))
         // { default = fattr default-ghc-version; };
@@ -43,8 +44,10 @@
                 hspkgs.cabal-install
                 hspkgs.haskell-language-server
                 hspkgs.hlint
-                hspkgs.ormolu
-                pkgs.bashInteractive # see: https://discourse.nixos.org/t/interactive-bash-with-nix-develop-flake/15486
+                # Ormolu doesn't work well with the way nix does old GHC
+                # versions, so I'm turning it off by default
+                # hspkgs.ormolu
+                pkgs.bashInteractive
                 pkgs.graphviz
               ];
             };

--- a/src/Calligraphy/Compat/Lib.hs
+++ b/src/Calligraphy/Compat/Lib.hs
@@ -18,10 +18,10 @@ module Calligraphy.Compat.Lib
   )
 where
 
+import qualified Calligraphy.Compat.GHC as GHC
 import Calligraphy.Util.Lens
 import Data.IORef
 import qualified Data.Set as Set
-import qualified Calligraphy.Compat.GHC as GHC
 
 #if MIN_VERSION_ghc(9,0,0)
 import GHC.Iface.Ext.Binary
@@ -50,7 +50,7 @@ getHieFiles filePaths = do
 getHieFiles filePaths = do
     uniqSupply <- GHC.mkSplitUniqSupply 'z'
     ref <- newIORef (GHC.initNameCache uniqSupply [])
-    forM hieFilePaths (readHieFileWithWarning ref)
+    forM filePaths (readHieFileWithWarning ref)
 
 #endif
 
@@ -63,7 +63,6 @@ readHieFileWithWarning ref path = do
     putStrLn $ "    This version of calligraphy was compiled with GHC version: " <> show GHC.hieVersion
     putStrLn "    Optimistically continuing anyway..."
   pure hie
-
 
 {-# INLINE sourceInfo #-}
 sourceInfo :: Traversal' (HieAST a) (NodeInfo a)

--- a/src/Calligraphy/Compat/Lib.hs
+++ b/src/Calligraphy/Compat/Lib.hs
@@ -22,6 +22,7 @@ import qualified Calligraphy.Compat.GHC as GHC
 import Calligraphy.Util.Lens
 import Data.IORef
 import qualified Data.Set as Set
+import Control.Monad
 
 #if MIN_VERSION_ghc(9,0,0)
 import GHC.Iface.Ext.Binary
@@ -30,7 +31,6 @@ import GHC.Types.Name.Cache
 import GHC.Types.SrcLoc
 import GHC.Utils.Outputable (ppr, showSDocUnsafe)
 import qualified Data.Map as Map
-import Control.Monad (forM, when)
 #else
 import HieBin
 import HieTypes

--- a/src/Calligraphy/Compat/Lib.hs
+++ b/src/Calligraphy/Compat/Lib.hs
@@ -14,12 +14,14 @@ module Calligraphy.Compat.Lib
     showAnns,
     mergeSpans,
     isPointSpan,
+    getHieFiles,
   )
 where
 
 import Calligraphy.Util.Lens
 import Data.IORef
 import qualified Data.Set as Set
+import qualified Calligraphy.Compat.GHC as GHC
 
 #if MIN_VERSION_ghc(9,0,0)
 import GHC.Iface.Ext.Binary
@@ -28,6 +30,7 @@ import GHC.Types.Name.Cache
 import GHC.Types.SrcLoc
 import GHC.Utils.Outputable (ppr, showSDocUnsafe)
 import qualified Data.Map as Map
+import Control.Monad (forM, when)
 #else
 import HieBin
 import HieTypes
@@ -35,11 +38,49 @@ import NameCache
 import SrcLoc
 #endif
 
+getHieFiles :: [FilePath] -> IO [HieFile]
+#if MIN_VERSION_ghc(9,4,0)
+
+getHieFiles filePaths = do
+  ref <- newIORef =<< GHC.initNameCache 'z' []
+  forM filePaths (readHieFileWithWarning ref)
+
+#else
+
+getHieFiles filePaths = do
+    uniqSupply <- GHC.mkSplitUniqSupply 'z'
+    ref <- newIORef (GHC.initNameCache uniqSupply [])
+    forM hieFilePaths (readHieFileWithWarning ref)
+
+#endif
+
+readHieFileWithWarning :: IORef GHC.NameCache -> FilePath -> IO GHC.HieFile
+readHieFileWithWarning ref path = do
+  GHC.HieFileResult fileHieVersion fileGHCVersion hie <- readHieFileCompat ref path
+  when (GHC.hieVersion /= fileHieVersion) $ do
+    putStrLn $ "WARNING: version mismatch in " <> path
+    putStrLn $ "    The hie files in this project were generated with GHC version: " <> show fileGHCVersion
+    putStrLn $ "    This version of calligraphy was compiled with GHC version: " <> show GHC.hieVersion
+    putStrLn "    Optimistically continuing anyway..."
+  pure hie
+
+
 {-# INLINE sourceInfo #-}
 sourceInfo :: Traversal' (HieAST a) (NodeInfo a)
 showContextInfo :: ContextInfo -> String
 readHieFileCompat :: IORef NameCache -> FilePath -> IO HieFileResult
-#if MIN_VERSION_ghc(9,0,0)
+
+#if MIN_VERSION_ghc(9,4,0)
+
+sourceInfo f (Node (SourcedNodeInfo inf) sp children) = (\inf' -> Node (SourcedNodeInfo inf') sp children) <$> Map.alterF (maybe (pure Nothing) (fmap Just . f)) SourceInfo inf
+
+showContextInfo = showSDocUnsafe . ppr
+
+readHieFileCompat ref path = do
+  nameCache <- readIORef ref
+  readHieFile nameCache path
+
+#elif MIN_VERSION_ghc(9,0,0)
 
 sourceInfo f (Node (SourcedNodeInfo inf) sp children) = (\inf' -> Node (SourcedNodeInfo inf') sp children) <$> Map.alterF (maybe (pure Nothing) (fmap Just . f)) SourceInfo inf
 

--- a/src/Calligraphy/Phases/Search.hs
+++ b/src/Calligraphy/Phases/Search.hs
@@ -8,10 +8,8 @@ module Calligraphy.Phases.Search
 where
 
 import qualified Calligraphy.Compat.GHC as GHC
-import Calligraphy.Compat.Lib (readHieFileCompat)
+import Calligraphy.Compat.Lib
 import Control.Applicative
-import Control.Monad.State
-import Data.IORef
 import Data.List (isPrefixOf)
 import Data.List.NonEmpty (NonEmpty (..), nonEmpty, toList)
 import Options.Applicative hiding (str)
@@ -20,12 +18,7 @@ import System.FilePath (isExtensionOf, (</>))
 
 searchFiles :: SearchConfig -> IO [GHC.HieFile]
 searchFiles SearchConfig {searchDotPaths, searchRoots, includePatterns, excludePatterns} = do
-  hieFilePaths <- searchHieFilePaths
-
-  hieFiles <- do
-    uniqSupply <- GHC.mkSplitUniqSupply 'z'
-    ref <- newIORef (GHC.initNameCache uniqSupply [])
-    forM hieFilePaths (readHieFileWithWarning ref)
+  hieFiles <- getHieFiles =<< searchHieFilePaths
 
   pure $
     flip filter hieFiles $ \file ->
@@ -50,16 +43,6 @@ searchFiles SearchConfig {searchDotPaths, searchRoots, includePatterns, excludeP
                   contents <- (if searchDotPaths then id else filter (not . isPrefixOf ".")) <$> listDirectory path
                   foldMap (go . (path </>)) contents
                 else pure []
-
-readHieFileWithWarning :: IORef GHC.NameCache -> FilePath -> IO GHC.HieFile
-readHieFileWithWarning ref path = do
-  GHC.HieFileResult fileHieVersion fileGHCVersion hie <- readHieFileCompat ref path
-  when (GHC.hieVersion /= fileHieVersion) $ do
-    putStrLn $ "WARNING: version mismatch in " <> path
-    putStrLn $ "    The hie files in this project were generated with GHC version: " <> show fileGHCVersion
-    putStrLn $ "    This version of calligraphy was compiled with GHC version: " <> show GHC.hieVersion
-    putStrLn $ "    Optimistically continuing anyway..."
-  pure hie
 
 data SearchConfig = SearchConfig
   { includePatterns :: Maybe (NonEmpty Pattern),

